src/cmd-cave.c

```diff
@@ -1340,6 +1339,264 @@ void do_cmd_run(struct command *cmd)
+		player->upkeep->update |= (PU_TORCH);
+		player->upkeep->running_withpathfind = true;
+		run_step(0);
+		return;
+	}
+
+	msg("No known path to downstairs.");
+}
+
+/**
+ * Automatically navigate to the nearest upstairs location.
+ *
+ * Note that navigating while confused is not allowed.
+ */
+void do_cmd_navigate_up(struct command *cmd)
+{
+	struct loc target_grid;
+	int visible_monster_count = 0;
+	/* cancel if confused */
+	if (player->timed[TMD_CONFUSED]) {
+		msg("You cannot explore while confused.");
+	   	return;
+	}
+
+
+	/* If we're in a web, deal with that */
+	if (square_iswebbed(cave, player->grid)) {
+		/* Clear the web, finish turn */
+		msg("You clear the web.");
+		square_destroy_trap(cave, player->grid);
+		player->upkeep->energy_use = z_info->move_energy;
+		return;
+	}
+	
+
+	int min_dist = 9999;
+	/* Get location */
+	for (int y = 0; y < cave->height; y++) {
+		for (int x = 0; x < cave->width; x++) {
+			struct loc grid = loc(x, y);
+			
+			if (loc_eq(grid, player->grid)) continue;
+
+			if (square_isoccupied(cave, grid)) {
+				int m_idx = square(cave, grid)->mon;
+				struct monster *mon = cave_monster(cave, m_idx);
+				if (monster_is_obvious(mon)) {
+					visible_monster_count++;
+					break;
+				}
+			}
+			if (square_isknown(cave, grid) && square_isupstairs(cave, grid)) {
+				int d = distance(grid, player->grid);
+				if (d < min_dist) {
+					target_grid = grid;
+					min_dist = d;
+				}
+			}
+		}
+	}
+
+	if (visible_monster_count > 0) {
+		msg("Something is here.");
+		return;
+	}
+
+	if (min_dist < 9999 && find_path(target_grid)) {
+		player->upkeep->running = 1000;
+		/* Calculate torch radius */
+		player->upkeep->update |= (PU_TORCH);
+		player->upkeep->running_withpathfind = true;
+		run_step(0);
+		return;
+	}
+
+	msg("No known path to upstairs.");
+}
+
+/**
+ * Start exploring.
+ *
+ * Note that exploring while confused is not allowed.
+ */
+void do_cmd_explore(struct command *cmd)
+{
+	struct loc target_grid;
+	bool visible_monster = false;
+	/* cancel if confused */
+	if (player->timed[TMD_CONFUSED]) {
+		msg("You cannot explore while confused.");
+	   	return;
+	}
+
+
+	/* If we're in a web, deal with that */
+	if (square_iswebbed(cave, player->grid)) {
+		/* Clear the web, finish turn */
+		msg("You clear the web.");
+		square_destroy_trap(cave, player->grid);
+		player->upkeep->energy_use = z_info->move_energy;
+		return;
+	}
+	
+
+	int min_dist = 9999;
+	/* Get location */
+	for (int y = 0; y < cave->height && !visible_monster; y++) {
+		for (int x = 0; x < cave->width; x++) {
+			struct loc grid = loc(x, y);
+			
+			if (loc_eq(grid, player->grid)) continue;
+
+			if (square_isoccupied(cave, grid)) {
+				int m_idx = square(cave, grid)->mon;
+				struct monster *mon = cave_monster(cave, m_idx);
+				if (monster_is_obvious(mon)) {
+					visible_monster = true;
+					break; /* only breaks the inner loop */
+				}
+			}
+			if (square_isnotknown(cave, grid) && square_ispassable(cave, grid)) {
```

Comment:
The autoexplore command leaks a bit of information here.  Something like square_isknown(cave, grid) && square_ispassable(cave, grid) && count_neighbors(NULL, c, grid, square_isnotknown, false) > 0 would not leak information but is more time consuming to evaluate and would cause autoexplore to pursue some more immediate dead ends.

Reply:
Ah, because this will know if the unexplored space is passable or not? Which the player usually would not know?
I see, you are right. Will be changed.